# 9 - Resumen de texto


<br>
<br>

<img src="https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/resumen.webp" style="width:400px;"/>

<h1>Tabla de Contenidos<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1---Modelos-de-resumen" data-toc-modified-id="1---Modelos-de-resumen-1">1 - Modelos de resumen</a></span></li><li><span><a href="#2---Pipeline-de-Transformers-para-resumir" data-toc-modified-id="2---Pipeline-de-Transformers-para-resumir-2">2 - Pipeline de Transformers para resumir</a></span></li><li><span><a href="#3---Usando-el-modelo-de-resumen" data-toc-modified-id="3---Usando-el-modelo-de-resumen-3">3 - Usando el modelo de resumen</a></span><ul class="toc-item"><li><span><a href="#3.1-Tokenizador" data-toc-modified-id="3.1-Tokenizador-3.1">3.1 Tokenizador</a></span></li><li><span><a href="#3.2---Modelo-de-resumen" data-toc-modified-id="3.2---Modelo-de-resumen-3.2">3.2 - Modelo de resumen</a></span></li><li><span><a href="#3.3---Resumen-funcional" data-toc-modified-id="3.3---Resumen-funcional-3.3">3.3 - Resumen funcional</a></span></li></ul></li><li><span><a href="#4---Más-modelos-de-resumen" data-toc-modified-id="4---Más-modelos-de-resumen-4">4 - Más modelos de resumen</a></span></li></ul></div>

## 1 - Modelos de resumen

Los modelos de resumen de texto son herramientas de NLP diseñadas para condensar información extensa en textos más cortos que mantienen los puntos clave del contenido original. Estos modelos pueden ser clasificados principalmente en dos tipos: extractivos y abstractivos.

1. **Resumen Extractivo**

Los modelos extractivos seleccionan y concatenan las frases o segmentos más importantes del texto original para formar un resumen. Estos modelos no generan nuevo texto, sino que extraen partes del texto original.

+ Técnicas comunes del resumen extractivo:
    + Análisis de frecuencia: Identifica las palabras y frases más frecuentes para determinar la importancia.
    + Algoritmos de clustering: Agrupan frases similares y seleccionan representantes de cada grupo.
    + Redes neuronales: Utilizan técnicas como LSTM o GRU para evaluar la importancia de las oraciones en el contexto del documento completo.


2. **Resumen Abstractivo**

Los modelos abstractivos generan un nuevo texto que no necesariamente aparece en el original, utilizando técnicas avanzadas para reformular y sintetizar contenido que refleje los puntos clave.

+ Técnicas comunes del resumen abstractivo:
    + Modelos Seq2Seq (Secuencia a Secuencia): Utilizan arquitecturas de codificador-decodificador, donde el texto original se codifica en una representación vectorial y el decodificador genera el resumen.
    + Transformers: Especialmente el modelo Transformer, que ha demostrado ser muy efectivo para resúmenes abstractivos debido a su capacidad para manejar largas dependencias en el texto.
    + BERT y derivados: Modelos como BERT, GPT y T5 han sido adaptados para mejorar las tareas de generación de texto, incluyendo el resumen abstractivo.



3. **Modelos populares de resumen**
    + GPT: Estos modelos de OpenAI, basados en la arquitectura Transformer, tienen la capacidad para realizar resúmenes abstractivos mediante el ajuste fino en tareas específicas de resumen.
    + BERTSUM: Una adaptación de BERT para resúmenes, que modifica la arquitectura BERT para hacerla más adecuada para seleccionar oraciones significativas.
    + T5 (Text-to-Text Transfer Transformer): Un modelo diseñado para convertir todas las tareas de NLP en un problema de generación de texto, incluido el resumen.


4. **Aplicaciones**
    + Medios de comunicación: Resumen de noticias y artículos para proporcionar versiones condensadas rápidamente.
    + Academia y educación: Resumen de documentos y literatura académica para facilitar la revisión y el estudio.
    + Negocios y legal: Automatización del resumen de informes, contratos y documentos legales para mejorar la eficiencia y reducir la carga de trabajo.
    
    
    
Los modelos de resumen no solo ofrecen eficiencias significativas en la gestión de la información, sino que también permiten a los usuarios acceder rápidamente a la información relevante sin necesidad de revisar grandes volúmenes de texto.

## 2 - Pipeline de Transformers para resumir

Vamos a usar un modelo llamado [mt5-small-spanish-summarization](https://huggingface.co/josmunpen/mt5-small-spanish-summarization). Este es un modelo mt5-small afinado para generar titulares a partir del cuerpo de las noticias en español. El modelo fue entrenado con 58.425 noticias extraídas de periódicos, 31.477 del periódico La Razón y 26.948 del periódico Público. Estas noticias pertenecen a las siguientes categorías: "España", "Cultura", "Economía", "Igualdad" y "Política".

In [1]:
from transformers import pipeline

In [2]:
tarea = 'summarization'

modelo = 'josmunpen/mt5-small-spanish-summarization'

In [3]:
resumen_pipe = pipeline(task=tarea, model=modelo)

/opt/anaconda3/envs/ia/lib/python3.9/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()
/opt/anaconda3/envs/ia/lib/python3.9/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
Hardware accelerator e.g. GPU is available in the environment, but no `device` argument is passed to the `Pipeline` object. Model will be on CPU.


In [4]:
# ejemplo de articulo

articulo = '''
La Guardia Civil ha desarticulado un grupo organizado dedicado a copiar en los examenes teoricos para 
la obtencion del permiso de conducir. Para ello, empleaban receptores y camaras de alta tecnologia y 
operaban desde la misma sede del Centro de examenes de la Direccion General de Trafico (DGT) en Mostoles. 
Es lo que han llamado la Operacion pinga. 
El grupo desarticulado ofrecia el servicio de transporte y tecnologia para copiar y 
poder aprobar. Por dicho servicio cobraban 1.000 euros. Los investigadores sorprendieron in fraganti a una 
mujer intentando copiar en el examen. Portaba una chaqueta con dispositivos electronicos ocultos, 
concretamente un telefono movil al que estaba conectada una camara que habia sido insertada en la parte 
frontal de la chaqueta para transmitir online el examen y que orientada al ordenador del Centro de Examenes 
en el que aparecen las preguntas, permitia visualizar las imagenes en otro ordenador alojado en el interior 
de un vehiculo estacionado en las inmediaciones del centro. En este vehiculo, se encontraban el resto del 
grupo desarticulado con varios ordenadores portatiles y tablets abiertos y conectados a paginas de test de la
DGT para consultar las respuestas. Estos, comunicaban con la mujer que estaba en el aula haciendo el examen 
a traves de un diminuto receptor bluetooth que portaba en el interior de su oido.  

Luis de Lama, portavoz de la Guardia Civil de Trafico destaca que los ciudadanos, eran de origen chino, 
y copiaban en el examen utilizando la tecnologia facilitada por una organizacion. Destaca que, ademas de parte 
del fraude que supone copiar en un examen muchos de estos ciudadanos desconocian el idioma, no hablan ni 
entienden el español lo que supone un grave riesgo para la seguridad vial por desconocer las señales y 
letreros que avisan en carretera de muchas incidencias. 

'''

In [7]:
len(articulo.split())

297

In [8]:
resumen = resumen_pipe(articulo, min_length=10, max_length=80)

resumen

[{'summary_text': 'la guardia civil desarticula un grupo dedicado a copiar en los examenes'}]

In [12]:
len(resumen[0]['summary_text'].split())

12

## 3 - Usando el modelo de resumen

Usemos el modelo fuera del pipeline. Como siempre, necesitamos el tokenizador y el modelo preentrenado:

In [13]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

### 3.1 Tokenizador

In [14]:
tokenizador = AutoTokenizer.from_pretrained(modelo)

In [15]:
tokenizador

T5TokenizerFast(name_or_path='josmunpen/mt5-small-spanish-summarization', vocab_size=250100, model_max_length=1000000000000000019884624838656, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'eos_token': '</s>', 'unk_token': '<unk>', 'pad_token': '<pad>'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("<pad>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	1: AddedToken("</s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	2: AddedToken("<unk>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}

Características principales del T5TokenizerFast:

1. **vocab_size**: 50265 representa el tamaño del vocabulario del tokenizador, es decir, la cantidad de tokens únicos que puede reconocer.


2. **model_max_length**: 1024 es la longitud máxima de tokens que el modelo puede manejar.


3. **is_fast**: True indica que este tokenizador utiliza la implementación rápida de Hugging Face, optimizada para ser más eficiente en comparación con las versiones basadas puramente en Python.


4. **padding_side y truncation_side**: Ambos configurados a right (derecha), lo que significa que el relleno y la truncación de secuencias se realizan al final de las secuencias de tokens.


5. **special_tokens**:Define varios tokens especiales que el modelo necesita para operar correctamente:
    + eos_token: `</s>` marca el fin de una secuencia.
    + unk_token: `<unk>` se utiliza para tokens desconocidos.
    + pad_token: `<pad>` se usa para rellenar secuencias hasta una longitud común.
    + mask_token: `<mask>` se usa para enmascarar y predecir la palabra original que ha sido enmascarada.


6. **clean_up_tokenization_spaces**: True indica que el tokenizador limpiará espacios de tokenización extras, ayudando a mantener la integridad del texto original en su conversión a tokens.



In [16]:
vector = tokenizador(articulo, return_tensors='pt')

In [19]:
vector.input_ids.shape

torch.Size([1, 474])

### 3.2 - Modelo de resumen

In [20]:
modelo_resumen = AutoModelForSeq2SeqLM.from_pretrained(modelo)

In [21]:
modelo_resumen

MT5ForConditionalGeneration(
  (shared): Embedding(250112, 512)
  (encoder): MT5Stack(
    (embed_tokens): Embedding(250112, 512)
    (block): ModuleList(
      (0): MT5Block(
        (layer): ModuleList(
          (0): MT5LayerSelfAttention(
            (SelfAttention): MT5Attention(
              (q): Linear(in_features=512, out_features=384, bias=False)
              (k): Linear(in_features=512, out_features=384, bias=False)
              (v): Linear(in_features=512, out_features=384, bias=False)
              (o): Linear(in_features=384, out_features=512, bias=False)
              (relative_attention_bias): Embedding(32, 6)
            )
            (layer_norm): MT5LayerNorm()
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (1): MT5LayerFF(
            (DenseReluDense): MT5DenseGatedActDense(
              (wi_0): Linear(in_features=512, out_features=1024, bias=False)
              (wi_1): Linear(in_features=512, out_features=1024, bias=False)
          

**Componentes del modelo MT5ForConditionalGeneration**:


1. **Shared Embedding**:

+ Embedding(250112, 512): Una capa de embedding compartida para el encoder y el decoder, con un vocabulario de 250,112 tokens y cada token representado por un vector de 512 dimensiones.


2. **Encoder (MT5Stack)**:

+ embed_tokens: Utiliza la misma capa de embedding compartida para transformar los tokens de entrada en vectores.
+ block: Una lista de módulos de tipo MT5Block, cada uno conteniendo capas de atención y capas de alimentación directa (feed-forward):
    + MT5LayerSelfAttention: Capa de auto-atención que ayuda al modelo a enfocarse en diferentes partes de la entrada para entender mejor el contexto.
    + MT5LayerFF (Feed Forward Layer): Capa que utiliza una activación no lineal (GELU) y operaciones lineales para transformar la salida de la capa de atención.


3. **Decoder (MT5Stack)**: Similar al encoder en estructura, pero con capas adicionales para la atención cruzada:

+ MT5LayerSelfAttention: Similar a la capa del encoder, se enfoca en la entrada del decoder.
+ MT5LayerCrossAttention (EncDecAttention): Permite que el decoder acceda a la salida completa del encoder, crucial para tareas como la traducción donde el contexto del texto fuente es esencial para generar el texto objetivo.
+ MT5LayerFF: Similar a la del encoder, procesa la salida de las capas de atención.


4. **Componentes Adicionales**:

+ Dropout(p=0.1): Utilizado en varias capas para evitar el sobreajuste mediante la desactivación aleatoria de neuronas durante el entrenamiento.
+ MT5LayerNorm: Normalización de capa para estabilizar las salidas internas del modelo.
+ lm_head: Una capa lineal que mapea la salida del decoder al espacio del vocabulario original, usada para generar la salida final del texto.


In [22]:
respuesta = modelo_resumen.generate(**vector)

/opt/anaconda3/envs/ia/lib/python3.9/site-packages/transformers/generation/utils.py:1258: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(


In [23]:
respuesta

tensor([[     0,    283,    259, 125617,   9242,    498, 144352,    262,    335,
           8237,    269,  94902,    259,    262,  57503,    286,    289,    595,
          56125,    299]])

In [26]:
tokenizador.batch_decode(respuesta, skip_special_tokens=True)[0]

'la guardia civil desarticula un grupo dedicado a copiar en los examenes'

### 3.3 - Resumen funcional
Vamos a poner todo el código junto en una sola función:

In [42]:
# librerias
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM



def resumen(texto: str, modelo: str) -> str:
    
    """
    Función para resumir un texto.
    
    Params:
    + texto: string. Texto a ser resumido.
    + modelo: string. Modelo de transformers para resumir.
    
    Return:
    String. Resumen.
    """
    
    # con este objeto vectorizamos las palabras
    tokenizador = AutoTokenizer.from_pretrained(modelo)
    
    
    # creación del vector
    vector = tokenizador(texto, return_tensors='pt')
    
    
    # inicializacion del modelo traductor
    modelo_traductor = AutoModelForSeq2SeqLM.from_pretrained(modelo)
    
    # tensor de respuesta del modelo
    respuesta = modelo_traductor.generate(**vector)
    
    # respuesta del modelo
    respuesta = tokenizador.batch_decode(respuesta, skip_special_tokens=True)[0]
    
    return respuesta

In [29]:
resumen(articulo, modelo)

'la guardia civil desarticula un grupo dedicado a copiar en los examenes'

## 4 - Más modelos de resumen 

Probemos ahora otros modelos, esta vez en inglés, puesto que la mayoría de modelos funcionan en este idioma. Vamos a usar un modelo de resumen de [Falconsai](https://huggingface.co/Falconsai/text_summarization), es un modelo T5 Small ajustado, una variante del modelo transformer T5, diseñado para la tarea de resumen de texto que pesa aproximadamente 250 Mb. Está adaptado y ajustado para generar resúmenes concisos y coherentes del texto de entrada. Este modelo está preentrenado en un corpus diverso de datos textuales, lo que le permite capturar información esencial y generar resúmenes significativos. El ajuste se realiza prestando especial atención a los ajustes de hiperparámetros, incluyendo el tamaño del lote y la tasa de aprendizaje, para garantizar un rendimiento óptimo en el resumen de texto.


El conjunto de datos de ajuste consta de una variedad de documentos y sus resúmenes generados por humanos correspondientes. Este conjunto de datos diverso permite al modelo aprender el arte de crear resúmenes que capturen la información más importante mientras mantienen la coherencia y la fluidez. El objetivo del proceso de entrenamiento es dotar al modelo de la capacidad de generar resúmenes de texto de alta calidad, haciéndolo valioso para una amplia gama de aplicaciones que involucran la resumición de documentos y la condensación de contenido.

In [31]:
# articulo de ejemplo

articulo = ''' 
Hugging Face: Revolutionizing Natural Language Processing
Introduction
In the rapidly evolving field of Natural Language Processing (NLP), Hugging Face has emerged as a prominent 
and innovative force. This article will explore the story and significance of Hugging Face, a company that has
made remarkable contributions to NLP and AI as a whole. From its inception to its role in democratizing AI, 
Hugging Face has left an indelible mark on the industry.
The Birth of Hugging Face
Hugging Face was founded in 2016 by Clément Delangue, Julien Chaumond, and Thomas Wolf. The name "Hugging Face" 
was chosen to reflect the company's mission of making AI models more accessible and friendly to humans, much 
like a comforting hug. Initially, they began as a chatbot company but later shifted their focus to NLP, 
driven by their belief in the transformative potential of this technology.
Transformative Innovations
Hugging Face is best known for its open-source contributions, particularly the "Transformers" library. 
This library has become the de facto standard for NLP and enables researchers, developers, and organizations 
to easily access and utilize state-of-the-art pre-trained language models, such as BERT, GPT-3, and more. 
These models have countless applications, from chatbots and virtual assistants to language translation and 
sentiment analysis.
Key Contributions:
1. **Transformers Library:** The Transformers library provides a unified interface for more than 50 
pre-trained models, simplifying the development of NLP applications. It allows users to fine-tune these 
models for specific tasks, making it accessible to a wider audience.
2. **Model Hub:** Hugging Face's Model Hub is a treasure trove of pre-trained models, making it simple for 
anyone to access, experiment with, and fine-tune models. Researchers and developers around the world can 
collaborate and share their models through this platform.
3. **Hugging Face Transformers Community:** Hugging Face has fostered a vibrant online community where 
developers, researchers, and AI enthusiasts can share their knowledge, code, and insights. This collaborative 
spirit has accelerated the growth of NLP.
Democratizing AI
Hugging Face's most significant impact has been the democratization of AI and NLP. Their commitment to 
open-source development has made powerful AI models accessible to individuals, startups, and established 
organizations. This approach contrasts with the traditional proprietary AI model market, which often limits 
access to those with substantial resources.
By providing open-source models and tools, Hugging Face has empowered a diverse array of users to innovate 
and create their own NLP applications. This shift has fostered inclusivity, allowing a broader range of voices 
to contribute to AI research and development.
Industry Adoption
The success and impact of Hugging Face are evident in its widespread adoption. Numerous companies and 
institutions, from startups to tech giants, leverage Hugging Face's technology for their AI applications. 
This includes industries as varied as healthcare, finance, and entertainment, showcasing the versatility 
of NLP and Hugging Face's contributions.
Future Directions
Hugging Face's journey is far from over. As of my last knowledge update in September 2021, the company was 
actively pursuing research into ethical AI, bias reduction in models, and more. Given their track record of 
innovation and commitment to the AI community, it is likely that they will continue to lead in ethical AI 
development and promote responsible use of NLP technologies.
Conclusion
Hugging Face's story is one of transformation, collaboration, and empowerment. Their open-source contributions 
have reshaped the NLP landscape and democratized access to AI. As they continue to push the boundaries of AI 
research, we can expect Hugging Face to remain at the forefront of innovation, contributing to a more 
inclusive and ethical AI future. Their journey reminds us that the power of open-source collaboration can 
lead to groundbreaking advancements in technology and bring AI within the reach of many.
'''

In [30]:
from transformers import pipeline

pipe = pipeline(task = "summarization", model="Falconsai/text_summarization")

Hardware accelerator e.g. GPU is available in the environment, but no `device` argument is passed to the `Pipeline` object. Model will be on CPU.


In [32]:
resumen_pipe = pipe(articulo)

Token indices sequence length is longer than the specified maximum sequence length for this model (851 > 512). Running this sequence through the model will result in indexing errors


In [35]:
resumen_pipe[0]['summary_text']

'Hugging Face has emerged as a prominent and innovative force in NLP . From its inception to its role in democratizing AI, the company has left an indelible mark on the industry . The name "Hugging Face" was chosen to reflect the company\'s mission of making AI models more accessible and friendly to humans .'

In [38]:
len(articulo.split())

608

In [39]:
len(resumen_pipe[0]['summary_text'].split())

55

Probemos otro modelo, un modelo [BART](https://huggingface.co/facebook/bart-large-cnn), preentrenado en inglés, ajustado con el dataset [CNN Daily Mail](https://huggingface.co/datasets/cnn_dailymail) y desarrollado por Facebook. Pesa 1.65Gb.

In [43]:
modelo = 'facebook/bart-large-cnn'

In [45]:
res = resumen(articulo, modelo)

In [46]:
res

"Hugging Face was founded in 2016 by Clément Delangue, Julien Chaumond, and Thomas Wolf. The company's most significant impact has been the democratization of AI and NLP. Their open-source contributions have reshaped the NLP landscape and democratized access to AI."

In [47]:
len(res.split())

40